src/script/common/c_content.cpp

```diff
@@ -574,14 +574,27 @@ void read_content_features(lua_State *L, ContentFeatures &f, int index)
 				break;
 			}
 		}
+
 		// Copy last value to all remaining textures
-		if(i >= 1){
-			TileDef lasttile = f.tiledef[i-1];
-			while(i < 6){
+		int max_i = 6;
+		if (f.drawtype == NDT_RAILLIKE)
+			max_i = 4; // special handling below
+		i = std::min(i, max_i);
+		if (i >= 1) {
+			const auto &lasttile = f.tiledef[i-1];
+			while (i < max_i) {
 				f.tiledef[i] = lasttile;
 				i++;
 			}
 		}
+
+		// Raillike: Fallback to 'straight' tile for 'single' and 'end' tile
+		if (f.drawtype == NDT_RAILLIKE) {
```

Comment:
What if the client is newer than the server? Shouldn't this code be client-side as well to avoid empty tile textures (index 4,5) on older servers?

Reply:
![end rail issue](https://user-images.githubusercontent.com/1497498/187026707-cbbf9c09-1ad6-4559-89d6-73d388e81588.png)

`ContentFeatures::updateTextures` should be the correct location for this code.

Reply:
This is tricky because the missing textures are already filled up by the time `updateTextures` runs. Filling up the tiles could be moved to the client but doesn't help with compatibility.

We need something like this either way:
* bump protocol version
* add code like this to `ContentFeatures::deSerialize`:
```js
if (protocol is old && drawtype is rail) {
    tiles[4] = tiles[0];
    tiles[5] = tiles[0];
}
```